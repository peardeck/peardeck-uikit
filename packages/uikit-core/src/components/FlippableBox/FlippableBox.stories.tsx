import React, { useState } from "react";
import { boolean, number, text } from "@storybook/addon-knobs";
import { IconImage } from "@peardeck/uikit-icons";
import { Box } from "../Box/Box";
import { Card } from "../Card/Card";
import { FlippableBox } from "./FlippableBox";
import { colorKnobs, spaceKnobs, themeColorKnob } from "../../theme/knobs";
import { Heading1 } from "../Heading/Heading";
import { Text } from "../Text/Text";
import { Button } from "../Button/Button";

export default {
  component: FlippableBox,
  title: "core|FlippableBox",
};

export const Default = () => (
  <Box {...colorKnobs()} {...spaceKnobs()} size={number("Container Size", 300)}>
    <FlippableBox
      front={
        <Box
          size={number("Front Face Size", 300)}
          bg={themeColorKnob("Front Face Background Color", "yellow")}
        >
          {text(
            "Front Face Content",
            "You are looking at the front of the card right now."
          )}
        </Box>
      }
      back={
        <Box
          size={number("Back Face Size", 300)}
          bg={themeColorKnob("Back Face Background Color", "blue")}
        >
          {text("Back Face Content", "This is the back!")}
        </Box>
      }
      flipped={boolean("Flipped", false)}
      flipSpeed={number("Flip Speed", 0.5)}
    />
  </Box>
);

const FlashcardComponent = ({ cardHeight }: { cardHeight: number }) => {
  const [flipped, setFlipped] = useState(false);

  return (
    <Box margin={4} width={[500, 700]}>
      <FlippableBox
        front={
          <Card
            height={cardHeight}
            bg="pear"
            display="flex"
            p={3}
            alignItems="stretch"
          >
            <Box
              display="flex"
              flex={1}
              mx={2}
              border="1px solid #698600"
              justifyContent="center"
              alignItems="center"
            >
              <IconImage height="3em" width="3em" />
            </Box>
            <Box mx={2} flex={2} p={2} overflow="auto">
              <Heading1>Definition</Heading1>
              <Text>
                {text(
                  "Definition",
                  "You are looking at the front of the card right now."
                )}
              </Text>
              <Heading1 pt={3}>Example</Heading1>
              <Text>{text("Example", "This is some example text. Blah")}</Text>
            </Box>
          </Card>
        }
        back={
          <Card
            height={cardHeight}
            bg="#eee"
            p={3}
            display="flex"
            alignItems="center"
            justifyContent="center"
          >
            <Heading1>{text("Term", "XXXXXXXXXXXXX")}</Heading1>
          </Card>
        }
        flipped={flipped}
        flipSpeed={number("Flip Speed", 0.5)}
      />
      <Box display="flex" justifyContent="center">
        <Button my={2} px={4} onClick={() => setFlipped(!flipped)}>
          Flip
        </Button>
      </Box>
    </Box>
  );
};

export const Flashcard = () => {
  const cardHeight = number("Card Height", 300);

  return <FlashcardComponent cardHeight={cardHeight} />;
};
