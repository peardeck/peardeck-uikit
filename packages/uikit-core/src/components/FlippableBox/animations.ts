import { keyframes } from "styled-components";

// Implementation that doesn't require explicit height/width
// shamelessly stolen from here: https://codepen.io/anon/pen/bwmkAo
// Added separate animation for front and back flips so that the
// animation "reverses" when flipped from back-to-front - that is,
// front-to-back flips clockwise, back-to-front flips counter-clockwise.

export const frontHide = keyframes`
  0% {
		transform: rotateY(0deg);
    height: 100%;
    width: 100%;
	}

	49% {
		height: 100%;
		width: 100%;
	}

	50% {
		height: 0;
		width: 0;
	}

	100% {
		transform: rotateY(180deg);
		height: 0;
		width: 0;
	}
`;

export const frontShow = keyframes`
  0% {
		transform: rotateY(180deg);
		height: 0;
		width: 0;
	}

	49% {
		height: 0;
		width: 0;
	}

	50% {
		height: 100%;
		width: 100%;
	}

	100% {
		transform: rotateY(0deg);
		height: 100%;
		width: 100%;
	}
`;

export const backHide = keyframes`
  0% {
		transform: rotateY(0deg);
    height: 100%;
    width: 100%;
	}

	49% {
		height: 100%;
		width: 100%;
	}

	50% {
		height: 0;
		width: 0;
	}

	100% {
		transform: rotateY(-180deg);
		height: 0;
		width: 0;
	}
`;

export const backShow = keyframes`
  0% {
		transform: rotateY(-180deg);
		height: 0;
		width: 0;
	}

	49% {
		height: 0;
		width: 0;
	}

	50% {
		height: 100%;
		width: 100%;
	}

	100% {
		transform: rotateY(0deg);
		height: 100%;
		width: 100%;
	}
`;
