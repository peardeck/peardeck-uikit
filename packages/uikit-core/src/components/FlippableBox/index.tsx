export { FlippableBox, FlippableBoxProps } from "./FlippableBox";
