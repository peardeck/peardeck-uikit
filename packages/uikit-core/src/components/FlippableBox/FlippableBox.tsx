import React, { useRef } from "react";
import styled, { Keyframes, CSSObject } from "styled-components";
import { Box } from "../Box/Box";
import { frontShow, frontHide, backShow, backHide } from "./animations";

interface FaceProps {
  shown: boolean;
  animate: boolean;
}

const Face = styled(Box)<FaceProps>`
  position: relative;
  /* overflow: hidden; */
`;

interface ContainerProps {
  flipSpeed: number;
}

const FlippableBoxContainer = styled(Box)<ContainerProps>`
  position: relative;
  perspective: 1000px;

  ${Face} {
    animation-duration: ${({ flipSpeed }): number => flipSpeed}s;
    animation-timing-function: linear;
    transform-style: preserve-3d;
    transition-property: transform;
    animation-fill-mode: forwards;
  }
`;

const FrontFace = styled(Face)`
  z-index: 2;
  transform: rotateY(0deg);
  animation-name: ${({ shown }): Keyframes => (shown ? frontShow : frontHide)};

  ${({ animate }): CSSObject =>
    animate ? {} : { height: "100%", width: "100%", animationName: "none" }}
`;

const BackFace = styled(Face)`
  z-index: 1;
  transform: rotateY(-180deg);
  animation-name: ${({ shown }): Keyframes => (shown ? backShow : backHide)};

  ${({ animate }): CSSObject =>
    animate
      ? {}
      : {
          height: "0",
          width: "0",
          visibility: "collapse",
          animationName: "none",
        }}
`;

export interface FlippableBoxProps extends Partial<ContainerProps> {
  front: React.ReactNode;
  back: React.ReactNode;
  flipped?: boolean;
}

export const FlippableBox = ({
  front,
  back,
  flipped = false,
  flipSpeed = 0.5,
}: FlippableBoxProps): JSX.Element => {
  // We don't want to animate the initial render of the element, so
  // once the element is mounted we can then use that knowledge to decide
  // whether we want to apply the keyframe animation.
  const containerRef = useRef(null);

  return (
    <FlippableBoxContainer flipSpeed={flipSpeed} ref={containerRef}>
      <FrontFace shown={!flipped} animate={!!containerRef.current}>
        {front}
      </FrontFace>
      <BackFace shown={flipped} animate={!!containerRef.current}>
        {back}
      </BackFace>
    </FlippableBoxContainer>
  );
};
