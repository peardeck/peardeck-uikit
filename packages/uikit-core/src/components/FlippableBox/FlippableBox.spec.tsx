import { create } from "react-test-renderer";
import React from "react";
import { FlippableBox } from "./FlippableBox";

describe("FlippableBox", () => {
  test("it matches the snapshot", () => {
    const json = create(
      <FlippableBox
        front={<div>This is the front</div>}
        back={<div>This is the back</div>}
      />
    ).toJSON();
    expect(json).toMatchSnapshot();
  });

  describe("flipped", () => {
    test("it matches the snapshot", () => {
      const json = create(
        <FlippableBox
          front={<div>This is the front</div>}
          back={<div>This is the back</div>}
          flipped
        />
      ).toJSON();
      expect(json).toMatchSnapshot();
    });
  });
});
