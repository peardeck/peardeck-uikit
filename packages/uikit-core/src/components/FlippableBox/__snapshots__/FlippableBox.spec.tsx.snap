// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlippableBox flipped it matches the snapshot 1`] = `
.c0 {
  height: 300px;
  width: 200px;
  -webkit-perspective: 1000px;
  -moz-perspective: 1000px;
  -ms-perspective: 1000px;
  perspective: 1000px;
}

.c1 {
  -webkit-transition-duration: 0.5s;
  transition-duration: 0.5s;
  -webkit-transform-style: preserve-3d;
  -ms-transform-style: preserve-3d;
  transform-style: preserve-3d;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  -webkit-transform: rotateY(180deg);
  -ms-transform: rotateY(180deg);
  transform: rotateY(180deg);
}

.c2 {
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  overflow: auto;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 2;
  -webkit-transform: rotateY(0deg);
  -ms-transform: rotateY(0deg);
  transform: rotateY(0deg);
}

.c3 {
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  overflow: auto;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 1;
  -webkit-transform: rotateY(180deg);
  -ms-transform: rotateY(180deg);
  transform: rotateY(180deg);
}

@media print {
  .c0 {
    -webkit-break-after: always;
    break-after: always;
  }
}

@media print {
  .c1 {
    -webkit-transform-origin: unset;
    -ms-transform-origin: unset;
    transform-origin: unset;
    -webkit-transition-duration: unset;
    transition-duration: unset;
    -webkit-transform: unset;
    -ms-transform: unset;
    transform: unset;
    position: unset;
    top: unset;
    bottom: unset;
    left: unset;
    right: unset;
  }
}

@media print {
  .c2 {
    -webkit-backface-visibility: unset;
    backface-visibility: unset;
    overflow: unset;
    position: unset;
    top: unset;
    bottom: unset;
    left: unset;
    right: unset;
    box-shadow: none;
    border-radius: 0;
    border: 1px dotted;
    border-bottom: 1px solid;
  }
}

@media print {
  .c2 {
    -webkit-transform: unset;
    -ms-transform: unset;
    transform: unset;
  }
}

@media print {
  .c3 {
    -webkit-backface-visibility: unset;
    backface-visibility: unset;
    overflow: unset;
    position: unset;
    top: unset;
    bottom: unset;
    left: unset;
    right: unset;
    box-shadow: none;
    border-radius: 0;
    border: 1px dotted;
    border-bottom: 1px solid;
  }
}

@media print {
  .c3 {
    -webkit-transform: rotate(180deg);
    -ms-transform: rotate(180deg);
    transform: rotate(180deg);
  }
}

<div
  className="sc-AxjAm c0"
  height={300}
  width={200}
>
  <div
    className="sc-AxjAm c1"
  >
    <div
      className="sc-AxjAm sc-AxhCb c2"
    >
      <div>
        This is the front
      </div>
    </div>
    <div
      className="sc-AxjAm sc-AxhCb c3"
    >
      <div>
        This is the back
      </div>
    </div>
  </div>
</div>
`;

exports[`FlippableBox it matches the snapshot 1`] = `
.c0 {
  height: 300px;
  width: 200px;
  -webkit-perspective: 1000px;
  -moz-perspective: 1000px;
  -ms-perspective: 1000px;
  perspective: 1000px;
}

.c1 {
  -webkit-transition-duration: 0.5s;
  transition-duration: 0.5s;
  -webkit-transform-style: preserve-3d;
  -ms-transform-style: preserve-3d;
  transform-style: preserve-3d;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
}

.c2 {
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  overflow: auto;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 2;
  -webkit-transform: rotateY(0deg);
  -ms-transform: rotateY(0deg);
  transform: rotateY(0deg);
}

.c3 {
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  overflow: auto;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 1;
  -webkit-transform: rotateY(180deg);
  -ms-transform: rotateY(180deg);
  transform: rotateY(180deg);
}

@media print {
  .c0 {
    -webkit-break-after: always;
    break-after: always;
  }
}

@media print {
  .c1 {
    -webkit-transform-origin: unset;
    -ms-transform-origin: unset;
    transform-origin: unset;
    -webkit-transition-duration: unset;
    transition-duration: unset;
    -webkit-transform: unset;
    -ms-transform: unset;
    transform: unset;
    position: unset;
    top: unset;
    bottom: unset;
    left: unset;
    right: unset;
  }
}

@media print {
  .c2 {
    -webkit-backface-visibility: unset;
    backface-visibility: unset;
    overflow: unset;
    position: unset;
    top: unset;
    bottom: unset;
    left: unset;
    right: unset;
    box-shadow: none;
    border-radius: 0;
    border: 1px dotted;
    border-bottom: 1px solid;
  }
}

@media print {
  .c2 {
    -webkit-transform: unset;
    -ms-transform: unset;
    transform: unset;
  }
}

@media print {
  .c3 {
    -webkit-backface-visibility: unset;
    backface-visibility: unset;
    overflow: unset;
    position: unset;
    top: unset;
    bottom: unset;
    left: unset;
    right: unset;
    box-shadow: none;
    border-radius: 0;
    border: 1px dotted;
    border-bottom: 1px solid;
  }
}

@media print {
  .c3 {
    -webkit-transform: rotate(180deg);
    -ms-transform: rotate(180deg);
    transform: rotate(180deg);
  }
}

<div
  className="sc-AxjAm c0"
  height={300}
  width={200}
>
  <div
    className="sc-AxjAm c1"
  >
    <div
      className="sc-AxjAm sc-AxhCb c2"
    >
      <div>
        This is the front
      </div>
    </div>
    <div
      className="sc-AxjAm sc-AxhCb c3"
    >
      <div>
        This is the back
      </div>
    </div>
  </div>
</div>
`;
